### ISSUE TITLE
PGML unmatching backticks in math mode causes hard to decipher warnings.

### ISSUE THREAD

**Issue by somiaj** [openwebwork/pg#654]
Found this issue when trying to debug a problem. When using PGML, if the number of starting and stopping ticks don't match, the warning makes it hard to determine what the issue is.

For example, ```[`\pi``]``` gives the following warning:

![image](https://user-images.githubusercontent.com/18082412/153331884-5b5ea735-4173-43f1-b402-2f419fba79d2.png)

While, ```[``\pi`]``` gives the following warning:

![image](https://user-images.githubusercontent.com/18082412/153331966-bce6860c-f5b6-4b8f-810e-9373ecb7d3f6.png)

It would be helpful if more useful warning messages were produced.

**Comment by Alex-Jordan** [comment]
And if you have unmatched colons like `[: pi ::]` you get no error (or output) at all.

You also get an error from unmatched pipes for comments: `[| pi ||]`.

**Comment by dpvc** [comment]
The intended result for an unmatched delimiter is that PGML issues a warning and treats the unmatched context as literal text.  That is what happens for me when I use my command-line tool for testing, so I was confused when that didn't happen in actual WeBWorK problems.

It took me a while to track it down, but the issue actually turns out to be fairly simple to fix, and a bit surprising.

Terminators come in two forms, literal strings an regex patterns, and there are a couple of places where `ref($block->{terminator})` is used to distinguish between the two.  PGML check if this `ref()` call equals `'Regexp'`, but somewhere along the line, the Safe compartment changed and, in the Safe compartment now, the `ref()` is `Safe::Root23::Regexp`, or something similar with different numbers.  That means that the test fails, and PGML goes down the wrong path for handling the missing ed-delimiter.  In this case, it leads to no PGML warning, but causes the log message you indicate (that turns out to be from passing an undefined math string to `EV3P()`).

The fix is to properly account for the new class type used in the Safe compartment, so that the right path will be taken, and the warning message about unmatched delimiters will generated via `WARN_MESSAGE()`.

Here is a diff that does that:

``` diff
diff --git a/macros/PGML.pl b/macros/PGML.pl
index d0a8756f..8610a58f 100644
--- a/macros/PGML.pl
+++ b/macros/PGML.pl
@@ -26,6 +26,11 @@ sub Eval {main::PG_restricted_eval(@_)}
 
 sub Sort {return main::lex_sort(@_)}
 
+sub isRegexp {
+  my $ref = shift;
+  return $ref && ref($ref) =~ m/Regexp$/;
+}
+
 ######################################################################
 
 package PGML::Parse;
@@ -84,7 +89,7 @@ sub Unwind {
   $self->{block}->popItem;
   $self->Text($block->{token});
   $self->{block}->pushItem(@{$block->{stack}});
-  $self->Text($block->{terminator}) if $block->{terminator} && ref($block->{terminator}) ne 'Regexp';
+  $self->Text($block->{terminator}) if $block->{terminator} && !PGML::isRegexp($block->{terminator});
   $self->{atBlockStart} = 0;
 }
 
@@ -192,7 +197,7 @@ sub End {
   my $block = $self->{block};
   $block->popItem if $block->topItem->{type} eq 'break' && $block->{type} ne 'align';
   while ($block->{type} ne 'root') {
-    if (ref($block->{terminator}) eq 'Regexp' || $block->{cancelPar}) {
+    if (PGML::isRegexp($block->{terminator}) || $block->{cancelPar}) {
       $self->blockError("'%s' was not closed before $action");
     } else {
       $self->Terminate;
@@ -965,13 +970,13 @@ sub Math {
     my $obj = Parser::Formula($context,$math);
     if ($context->{error}{flag}) {
       PGML::Warning "Error parsing mathematics: $context->{error}{message}";
-      return "\\text{math error}";
+      return ("\\text{math error}", "inline");
     }
     $obj = $obj->reduce if $item->{reduced};
     $math = $obj->TeX;
   }
   $math = "\\displaystyle{$math}" if $item->{displaystyle};
-  my $mathmode = ($item->{display}) ? 'display' : 'inline' ;
+  my $mathmode = ($item->{display} ? 'display' : 'inline');
   return ($math,$mathmode);
 }
```

[This is a diff against the 2.16 code-base, which is the highest version I currently have running.  I haven't ben able to get 2.18 running, and haven't taken time to figure out what is going wrong.  I also can't push changes to by clone of PG, as the workflows die and I haven't tried to figure that out either.  Sorry for not being able to do this properly.]

Here the `PGML::isRegexp()` function does the checking for a regular expression by testing if the `ref()` ends in `Regexp`.  I think that is sufficient, as there are no other objects that `block->{terminator}` can be, but you could match against `m/^(?:Safe::.*)?Regexp$/` if you want to be more definitive, or even `m/^(?:Safe::Root\d+::)?Regexp$/`

There is an argument that ```[`\pi``]``` should end the ``[` `` with the math being ``\pi` ``, but since you can do ``[`\pi` `]`` to get that, I don't think it is work trying to change the processing of the end delimiter at this point.  \

In any case, this should now get you the message

```
---- PGML (eval 3713) 1505 ------
==================
Errors parsing PGML:
'[`' was not closed before END_PGML
==================
__________________________
```

in the warning area, and the output should be

```
[`\pi``]
```

literally.  (But because MathJax runs in the page, it will pick up the `` `\pi` `` as AsciiMath and typeset that.  But I think one of the other recent PGML changes will prevent that.)

**Comment by Alex-Jordan** [comment]
This is fixed by #980.
